softlayer/resource_softlayer_network_vlan.go

```diff
@@ -0,0 +1,419 @@
+	sess := meta.(*session.Session)
+	service := services.GetNetworkVlanService(sess)
+
+	vlanId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid vlan ID, must be an integer: %s", err)
+	}
+
+	vlan, err := service.Id(vlanId).Mask(
+		"id," +
+			"name," +
+			"primaryRouter[datacenter[name]]," +
+			"primaryRouter[hostname]," +
+			"vlanNumber," +
+			"billingItem[recurringFee]," +
+			"guestNetworkComponentCount," +
+			"subnets[networkIdentifier,cidr,subnetType]",
+	).GetObject()
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving Network Vlan: %s", err)
+	}
+
+	d.Set("id", *vlan.Id)
+	d.Set("vlan_number", *vlan.VlanNumber)
+	d.Set("child_resource_count", *vlan.GuestNetworkComponentCount)
+	if vlan.Name != nil {
+		d.Set("name", *vlan.Name)
+	} else {
+		d.Set("name", "")
+	}
+
+	if vlan.PrimaryRouter != nil {
+		d.Set("primary_router_hostname", *vlan.PrimaryRouter.Hostname)
+		if strings.HasPrefix(*vlan.PrimaryRouter.Hostname, "fcr") {
+			d.Set("type", "PUBLIC")
+		} else {
+			d.Set("type", "PRIVATE")
+		}
+		if vlan.PrimaryRouter.Datacenter != nil {
+			d.Set("datacenter", *vlan.PrimaryRouter.Datacenter.Name)
+		}
+	}
+
+	if vlan.BillingItem == nil {
+		d.Set("softlayer_managed", true)
+	} else {
+		d.Set("softlayer_managed", false)
+	}
+
+	// Subnets
+	subnets := make([]map[string]interface{}, 0)
+
+	for _, elem := range vlan.Subnets {
+		subnet := make(map[string]interface{})
+		subnet["subnet"] = *elem.NetworkIdentifier + "/" + strconv.Itoa(*elem.Cidr)
+		subnet["subnet_type"] = *elem.SubnetType
+		subnets = append(subnets, subnet)
+	}
+	d.Set("subnets", subnets)
+
+	if vlan.Subnets != nil && len(vlan.Subnets) > 0 {
+		d.Set("primary_subnet_size", 1<<(uint)(32-*vlan.Subnets[0].Cidr))
+	} else {
+		d.Set("primary_subnet_size", 0)
+	}
+
+	return nil
+}
+
+func resourceSoftLayerNetworkVlanUpdate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(*session.Session)
+	service := services.GetNetworkVlanService(sess)
+
+	vlanId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid vlan ID, must be an integer: %s", err)
+	}
+
+	opts := datatypes.Network_Vlan{}
+
+	if d.HasChange("name") {
+		opts.Name = sl.String(d.Get("name").(string))
+	}
+
+	_, err = service.Id(vlanId).EditObject(&opts)
+
+	if err != nil {
+		return fmt.Errorf("Error updating Network Vlan: %s", err)
+	}
+	return resourceSoftLayerNetworkVlanRead(d, meta)
+}
+
+func resourceSoftLayerNetworkVlanDelete(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(*session.Session)
+	service := services.GetNetworkVlanService(sess)
+
+	vlanId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid vlan ID, must be an integer: %s", err)
+	}
+
+	// Check if the VLAN is existed.
+	_, err = service.Id(vlanId).Mask("id").GetObject()
+	if err != nil {
+		return fmt.Errorf("Error deleting Network Vlan: %s", err)
+	}
```

Comment:
Don't think we need to check if the vlan existed separately. I think the next call to the api below will tell you that already. If it existed, this results in two api calls, rather than just one.